test/e2e/e2e.go

```diff
@@ -0,0 +1,289 @@
+package e2e
+
+import (
+	"fmt"
+	"strings"
+	"testing"
+	"time"
+
+	channelsV1alpha1 "github.com/knative/eventing/pkg/apis/channels/v1alpha1"
+	feedsV1alpha1 "github.com/knative/eventing/pkg/apis/feeds/v1alpha1"
+	flowsV1alpha1 "github.com/knative/eventing/pkg/apis/flows/v1alpha1"
+	"github.com/knative/eventing/test"
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	rbacV1beta1 "k8s.io/api/rbac/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+
+	// Mysteriously required to support GCP auth (required by k8s libs).
+	// Apparently just importing it is enough. @_@ side effects @_@.
+	// https://github.com/kubernetes/client-go/issues/242
+	_ "k8s.io/client-go/plugin/pkg/client/auth/gcp"
+)
+
+const (
+	defaultNamespaceName = "e2etestfn3"
+	testNamespace        = "e2etest"
+	interval             = 1 * time.Second
+	timeout              = 1 * time.Minute
+)
+
+// Setup creates the client objects needed in the e2e tests.
+func Setup(t *testing.T, logger *zap.SugaredLogger) (*test.Clients, *test.Cleaner) {
+	if test.Flags.Namespace == "" {
+		test.Flags.Namespace = defaultNamespaceName
+	}
+
+	clients, err := test.NewClients(
+		test.Flags.Kubeconfig,
+		test.Flags.Cluster,
+		test.Flags.Namespace)
+	if err != nil {
+		t.Fatalf("Couldn't initialize clients: %v", err)
+	}
+	cleaner := test.NewCleaner(logger, clients)
+
+	return clients, cleaner
+}
+
+// TearDown will delete created names using clients.
+func TearDown(clients *test.Clients, cleaner *test.Cleaner, logger *zap.SugaredLogger) {
+	cleaner.Clean()
+
+	// There seems to be an Istio bug where if we delete / create
+	// VirtualServices too quickly we will hit pro-longed "No health
+	// upstream" causing timeouts.  Adding this small sleep to
+	// sidestep the issue.
+	//
+	// TODO(#1376):  Fix this when upstream fix is released.
+	logger.Info("Sleeping for 20 seconds after clean to avoid hitting issue in #1376")
+	time.Sleep(20 * time.Second)
+}
+
+// CreateRouteAndConfig will create Route and Config objects using clients.
+// The Config object will serve requests to a container started from the image at imagePath.
+func CreateRouteAndConfig(clients *test.Clients, logger *zap.SugaredLogger, cleaner *test.Cleaner, name string, imagePath string) error {
```

Comment:
It is similar but slightly different.

Reply:
@srinivashegde86 can we make this flexible and have a single instance in knative/pkg?

Reply:
There are two reasons why this is now in the eventing repo.
1) I'm using the `Cleaner` and adding the Route and Config to this Cleaner so that is can be later deleted. The point is deleting the objects in reverse order. This is currently not done in knative/serving and I had problems when deleting some resources because a ServiceAccount was deleted earlier and Feed/EventType could never be deleted afterwards. So maybe we could move the Cleaner to `knative/pkg` and use it in `knative/serving` as well
2) it is more flexible to leave creation of some objects in the individual repos. Having this in `pkg` means they have to be more general. And if they're not general enough in the future they will require changes in 3 repos (pkg, serving, eventing) rather than just one. The objects are too specific to be moved to pkq, IMO. My 2c

Reply:
I think [1] is a very compelling argument. Can you please open an issue?
I get your point in [2], I was suggesting just moving the common stuff to `pkg`.
But filing the issue would be enough, so I see no reason to block this PR.
Thanks!

Reply:
Cool, created https://github.com/knative/eventing/issues/431